drivers: pwm: nrf: nrf pwm driver modification to use DT

gamnes opened the pull request:
1. Kconfig options clock_prescaler removed.
2. Kconfig options CH pin selection moved to DT.
3. pwm device node added with alias to all
   nRF5x DTSI files.
4. Added yaml binding for Nordic PWM node.
5. Rewritten pwm_nrfx.c nRF HW PWM driver.
6. Modified fade_led PWM example to also
   include nRF HW PWM option.

Fixes #10035

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes commented:
This pull request is done as part of issue [8758](https://github.com/zephyrproject-rtos/zephyr/issues/8758). 

This pull request fixes issue [10035](https://github.com/zephyrproject-rtos/zephyr/issues/10035).

This pull request assumes issue [6958](https://github.com/zephyrproject-rtos/zephyr/issues/6958) will be merged and function get_cycles_per_sec() will eventually be removed.

Setting duty to either 0% or 100% for all channels of a PWM instance will disable the peripheral, see [9507](https://github.com/zephyrproject-rtos/zephyr/issues/9507) - if better semantics are introduced, this should be fixed in this driver as well.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -10,88 +10,221 @@
-	nrfx_pwm_t pwm;
-	nrfx_pwm_config_t config;
-	nrf_pwm_sequence_t seq;
+	NRF_PWM_Type * pwm_regs;                            // PWM Instance
```
I don't entirely understand the need of completely rewriting the driver and no longer using nrfx as the underlying layer. Could you elaborate on this a bit?

gamnes replied:
As discussed; will be reverting to using the nrfx layer as reasons for rewrite are no longer implemented.

gamnes commented on dts/bindings/pwm/nordic,nrf-pwm.yaml:
```diff
@@ -0,0 +1,71 @@
+version: 0.1
+
+description: >
+    This binding gives a base representation of the NRF5 PWM
```
(Y)

gamnes commented on dts/bindings/pwm/nordic,nrf-pwm.yaml:
```diff
@@ -0,0 +1,71 @@
+    ch0-pin:
+      type: int
+      description: Channel 0 pin
+      category: required
```
Done. Is this enforced by the tools in any way? If required - will a build where ch0-pin is not set fail?

ioannisg reviewed (dismissed):
Why is only the nrf52_pca10040 board DTS fixed?

carlescufi commented:
@ioannisg @anangl can you re-review please?

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+		if (matching_prescaler_found == false) {
+			LOG_ERR(
+			"Matching prescaler for period_cycles %d not found.\n",
+			period_cycles);
```
Such indentation is misleading. Consider shortening the error message.

gamnes replied:
Done.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+	 */
+	ret = pwm_period_check_and_set(pdata, pwm_instance, pwm, period_cycles);
+	if (ret) {
+		LOG_ERR("Incompatible period %d, already in use: %d",
```
This message will be shown also when the instance is not in use but there is no suitable prescaler available. In such case "already in use: x" will be only misleading.

gamnes replied:
Agreed. I will remove the already in use comment.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+	}
+
+	/* Check if all channels are off (duty 0% or 100%) */
+	if (!pwm_channel_active(pdata)) {
```
Wouldn't `any_channel_active` work better?

gamnes replied:
Agreed.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+		}
+
+		/* Start sequence */
+		nrfx_pwm_simple_playback(&pconfig->pwm,
```
This should be inside the `if` block above.

gamnes replied:
This has to be called anytime we have a new sequence to play. So it needs to be outside the if block.

If pwm is not in use, start it. Then start new playback.
If pwm is in use, we simply have to start a new playback to get the new values.

I agree that this is confusing, i will add some comments to make it more clear what is being done.

gamnes replied:
Right, thanks. Now i understood what you mean. The nrfx layer is using both sequences and connecting them through the use of shorts, it will keep loading in new values all the time from the same array and thus we only have to start the sequence when its not already running. Will fix it.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -126,8 +288,10 @@ static int pwm_nrfx_init(struct device *dev)
 static void pwm_nrfx_uninit(struct device *dev)
 {
 	const struct pwm_nrfx_config *pconfig = dev->config->config_info;
+	NRF_PWM_Type *pwm_instance = pconfig->pwm_regs;
```
No such field in the `pwm_nrfx_config` structure.

gamnes replied:
Thanks! Did not see this one as i have not tested this function. Will make it correct

anangl commented:
Split these changes into several commits. Use separate commits for logically related changes (for instance, don't add DTS entries and change PWM prescaler handling in one commit as these things are unrelated).

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
 
 struct pwm_nrfx_data {
+	nrfx_pwm_config_t config;
+	nrf_pwm_sequence_t seq;
```
The config values are being edited in pwm_period_check_and_set() function where a new base_clock, top_value, etc, is being selected based on the input period_cycles.

anangl replied:
Some configuration values are changed, indeed. But those are contained in `nrfx_pwm_config_t`, not in `nrf_pwm_sequence_t`. The latter structure stays untouched. Only the values _it points to_ are changed. This also implies that there is no point in restarting the playback (what I tried to point out by making [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/11726#discussion_r238638359)).

gamnes replied:
As discussed, will be moving back the config and seq structures to pwm_nrfx_config and create pointers in pwm_nrfx_data to the values we want to change.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
 };
 
 struct pwm_nrfx_data {
+	nrfx_pwm_config_t config;
```
That can of course be done if desired. I just feel like placing this structure under pwm_nrfx_config, and then using pwm_nrfx_config as a const struct under DEVICE_DEFINE might be confusing since we in fact are changing some of the members in it when its used; but i have no strong opinions on this, if you want me to change it back and add pointers in the data struct i can do that.

Another point i would like to make is that i foresee us being able to change the PWM pins in the future. There is nothing stopping us from changing the PWM output GPIOs in runtime when none of the channels are in use; so that leaves even more of the config struct changing at run-time?

gamnes replied:
As discussed, moved configuration back to config struct.

gamnes commented on dts/bindings/pwm/nordic,nrf-pwm.yaml:
```diff
@@ -0,0 +1,71 @@
+      generation: define
+
+    ch0-pin-inverted:
+      type: int
```
Agreed. Will change it.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
 };
 
+static u32_t pwm_period_check_and_set(
+	struct pwm_nrfx_data *pdata, NRF_PWM_Type *pInstance,
```
Thanks. Was left behind from my previous version. Fixed now. No camel-case used.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+	nrf_pwm_sequence_t seq;
 	u16_t current[NRF_PWM_CHANNEL_COUNT];
-	u16_t top_value;
+	u16_t pwm_instance_in_use;
```
No reason. Assumed bool was not used. Will change to use bool instead.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+		 */
+		for (int i = 0; i <= PWM_PRESCALER_PRESCALER_Msk; i++) {
+			u16_t divisor = 1 << i;
+			u32_t new_countertop = period_cycles / divisor;
```
Changed it to use config.base_clock, but could of course add it back

gamnes commented:
Ok. Will split these changes into multiple commits.

galak commented on dts/bindings/pwm/nordic,nrf-pwm.yaml:
```diff
@@ -0,0 +1,71 @@
+      category: optional
+      generation: define
+
+    ch1-pin-inverted:
```
chX-pin-inverted

This should be a boolean type.

gamnes replied:
Done.

galak reviewed (dismissed):
one DT change to the binding.

gamnes committed:
dts: nrf: PWM device node added to nRF52 devices with yaml binding

1. PWM device node added with alias to all
   nRF52x DTSI files. 1 instance for
   nRF52810, 3 instances for nRF52832, and
   4 instance for nRF52840.
2. Added yaml binding for Nordic PWM node.

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes committed:
drivers: pwm: nrf: Migrate Kconfig options to DT

1. Kconfig options CH pin removed.
2. Kconfig options CH inverted removed.
3. Modified pwm_nrfx.c driver to use DT
   defines instead of Kconfig.

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes committed:
drivers: pwm: nrf: Removed Kconfig prescaler option. Rewrite HW driver

1. Kconfig options clock_prescaler removed.
2. Rewritten pwm_nrfx.c nRF HW PWM driver.

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes committed:
drivers: pwm: nrf: Modify fade_led PWM example, use new HW drv

1. Modified fade_led PWM example to include
   nRF HW PWM option.
2. Added fade_led nrf52_pca10040.overlay
   in order to enable PWM node and choose
   output PWM GPIO for channel 0. Channel 1
   GPIO enable but not used in sample src.

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes commented:
@anangl: Split up into multiple commits, and all the changes you requested should have been resolved.
@galak: DT changed to used boolean instead of int.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
+	if (!any_channel_active(pdata)) {
+		/* If peripheral in use and all channels off: disable. */
+		if (pdata->pwm_instance_in_use) {
+			nrf_pwm_disable(pwm_instance);
```
The sequence should be stopped first to avoid any glitch (it depends on the used period length, but it is possible at this point that the new value written to `current` is not effective yet - the old one is still played; and `nrf_pwm_disable` acts immediately).
Actually, I wonder if `nrfx_pwm_stop` wouldn't be enough. Perhaps there is no need to disable the peripheral completely (this will be anyway done by power management when a low power state is entered).

gamnes replied:
As discussed, nrfx_pwm_stop() would be enough.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -126,8 +294,10 @@ static int pwm_nrfx_init(struct device *dev)
 	const struct pwm_nrfx_config *pconfig = dev->config->config_info;
+	NRF_PWM_Type *pwm_instance = pconfig->pwm.p_registers;
 
-	nrfx_pwm_uninit(&pconfig->pwm);
```
Why? The driver has to be deinitialized when a low power state is entered. Otherwise, the re-initialization when returning to the active state will fail.

gamnes replied:
Agreed.

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
+	return (current & NRFX_PWM_COMPARE_MASK);
+}
+
+static bool get_pin_inversion_state(u16_t config_output_pin)
```
`u8_t config_output_pin`
Do we actually need a separate function for this? It is used only in one place and its result is assigned to a variable named `channel_inverted_state` so the masking could be performed directly there, and the code will still be readable (and even more clearly showing what is actually going on).

gamnes replied:
Removed.

anangl commented:
> @anangl: Split up into multiple commits, and all the changes you requested should have been resolved.

The changes are split but all the commits have the same title, which is not always adequate to what is actually changed by the commit (for instance the first commit add only dts entries and bindings, but the title says "drivers: pwm: nrf: ...").

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
 
+static u32_t pwm_period_check_and_set(const struct pwm_nrfx_config *pconfig,
+	struct pwm_nrfx_data *pdata, NRF_PWM_Type *pwm_instance, u32_t pwm,
+	u32_t period_cycles)
```
Alright.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
+		/* If period_cycles fit with standard prescaler,
+		 * set it directly
+		 */
+		pdata->prescaler = PWM_PRESCALER_PRESCALER_DIV_1;
```
Ok.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
+	return 0;
+}
+
+static u8_t pwm_channel_map(const struct pwm_nrfx_config *pconfig, u32_t pwm)
```
True. Will change

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
-	if (pulse_cycles > pdata->top_value) {
+	/* Check if pulse is bigger than period, fail if so */
+	if (pulse_cycles > period_cycles) {
 		return -EINVAL;
```
Done.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
 #define PWM_NRFX_CH_VALUE_NORMAL        (1UL << 15)
 #define PWM_NRFX_CH_VALUE_INVERTED      (0)
+#define NRFX_PWM_PIN_INVERTED_MASK      0x80
+#define NRFX_PWM_PIN_MASK               0x7F
```
OK. I will look at how to change this, but i will not be using ~NRFX...INVERTED for the PIN mask as that looks cryptic to me.

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
+#define NRFX_PWM_PIN_INVERTED_MASK      0x80
+#define NRFX_PWM_PIN_MASK               0x7F
+#define NRFX_PWM_POLARITY_MSK           0x8000
+#define NRFX_PWM_COMPARE_MASK           0x7FFF
```
Looks better at least (Y)

gamnes commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,258 @@
+		 */
+		for (u8_t prescaler = 0;
+				prescaler <= PWM_PRESCALER_PRESCALER_Msk;
+				prescaler++) {
```
Done.

gamnes commented:
Got it, will change the wording of the commit message to be more clear

gamnes commented:
@anangl - Please see if changes were OK. Thanks

anangl commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -198,7 +334,7 @@ static int pwm_nrfx_pm_control(struct device *dev,
 #define PWM_NRFX_OUTPUT_PIN(dev_idx, ch_idx)		      \
 	(DT_NORDIC_NRF_PWM_PWM_##dev_idx##_CH##ch_idx##_PIN | \
-	 (DT_NORDIC_NRF_PWM_PWM_##dev_idx##_CH##ch_idx##_INVERTED ?   \
+	(DT_NORDIC_NRF_PWM_PWM_##dev_idx##_CH##ch_idx##_INVERTED ?   \
```
This was okay.  The second line should be indented with one space because its opening parenthesis is the inner one. 
What could be corrected here instead is the alignment (actually, lack of it) of the `\` characters.

gamnes replied:
(Y)

anangl reviewed (commented):
Looks good.
Two minor corrections suggested.

gamnes commented on drivers/pwm/Kconfig.nrfx:
```diff
@@ -1,296 +1,79 @@
-# Copyright (c) 2018, Cue Health Inc
```
Yes, will look into it

gamnes replied:
I think there was one LF line feed in the middle of the file that made this happen. Changed that one line and fix all the commits. Force pushed new. Lets see if that worked.

gamnes commented:
Thanks @anangl - The last force-push was just to align some indentation in pwm_nrfx.c

Mierunski commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,236 @@
 };
 
+static u32_t pwm_period_check_and_set(const struct pwm_nrfx_config *pconfig,
+				      struct pwm_nrfx_data *pdata,
```
If 'p' in pdata and pconfig stand for pointer it should be [removed.](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#naming)

gamnes replied:
Done. Removed all p prefix on variable names.

Mierunski commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,236 @@
+	nrf_pwm_configure(pwm_instance,
+		pdata->prescaler,
+		pconfig->initial_config.count_mode,
+		pdata->countertop);
```
Please align these lines with first argument to function.

gamnes replied:
done. As mentioned earlier, also removed p prefix on variable names.

Mierunski commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -5,93 +5,236 @@
 static int pwm_nrfx_pin_set(struct device *dev, u32_t pwm,
 			    u32_t period_cycles, u32_t pulse_cycles)
 {
+	/* We assume here that period_cycles will always be 16NHz
```
16Mhz

gamnes replied:
Done.

Mierunski commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -104,20 +247,13 @@ static int pwm_nrfx_init(struct device *dev)
-					  NULL);
-
+	nrfx_err_t result = nrfx_pwm_init(&pconfig->pwm,
+					  &pconfig->initial_config, NULL);
```
Move last argument to new line

gamnes replied:
Done.

carlescufi commented:
@gamnes can you fix CI?

Mierunski reviewed (changes requested):
Looks good, just a few nitpicks

carlescufi commented:
@Mierunski can you take another look?

gamnes commented:
Latest force-push was rebase onto upstream master.
